Add unit testing (and rework the sensors a bit)

DrewBrandt opened the pull request:
# Summary:

This PR does a lot of restructuring and somewhat significantly modifies the behavior of the sensors.

I've added numerous (like, 50) unit tests for various parts of the base sensor classes, the state class, and the APRS encoding. These tests are only designed to be used on PlatformIO at the moment, and uses its structure for testing. They rely on the [**Unity**](https://github.com/ThrowTheSwitch/Unity) testing framework which, according to Unity, is mostly just a collection of easy-to-run asserts. I don't know enough about how the Arduino IDE handles testing to know if this would compile properly on it. I did try to make sure that the library being used in the Arduino IDE would compile properly, though. Further testing for this is required.

Tests currently include:
- APRS messaging
- GPS functionality
- Barometer functionality
- Matrix
- Circular Buffer

More testing documentation to follow.

# Changes:

### Major:
- Added a mock Arduino framework for use in testing when using the `native`, or desktop, environment. Everything for the `native` testing environment is located in either `./test/` or `./lib`. This should allow Arduino IDE to properly compile the library using only `./src/` Arduino functions mocked include:
  - delay()
  - digitalWrite()
  - millis()
  - the Serial class (write, print, println)
  - SdFat (with FsFile and SdFs)
  - HIGH and LOW pin values
  - INPUT and OUTPUT pin modes
- Added fake sensors designed to test the base implementation of each sensor type. Includes an IMU, GPS, and Barometer.
- Added a `NativeTestHelper.h` that sets up common test variables and included files.
- Added a `native` environment to the `platformio.ini` to allow for testing on a desktop.
- Reworked the way sensors should be initialized. Call `[SensorType].begin()` instead of `[SensorType].initialize()` Full reasoning will be made clear when more documentation is written.
- Added a circular buffer type to the `Math` folder. Designed to be used with the sensors for storing calibration data, but can be used by anyone to hold anything. Documentation on how to use will come later.
- Removed all references to `imumaths` in favor of our own `mmfs::vector`, with the sole exception of the BNO hardware classes, where a conversion method is included.
- Added conditional includes for many places where `<Arduino>` is included, replacing it with a condition to possibly include the mock version if necessary.
- Introduced a calibration buffer to Barometer and GPS. These optional buffers consistently average the latest data to keep the "origin" point of their readings to 0, essentially combating drift. Call `Sensor.markLiftoff()` to save the latest value and stop calibration. More documentation to come later.
- Removed the unused time functions from the GPS. If it needs to be re-implemented, a complete re-write would be best anyways.
- Changed GPS's `pos` vector of size 2 to a `position` vector of size 3, adding altitude to the `z` component.
- Changed GPS's distance calculation away from a flat-earth approximation to a faster approximation of Vincenty's Formulae. Taken from [this](https://blog.mapbox.com/fast-geodesic-approximations-with-cheap-ruler-106f229ad016) article and [this](https://github.com/mapbox/cheap-ruler/blob/main/index.js#L475) repo. (Supersedes #29 with better approximation)
- Renamed the `ERROR` LogType  to `ERROR_` because apparently Windows decided to make `ERROR` a macro :/


### Minor:
- Moved `TestingImplementation` folder into its own `Examples` folder outside of `src`.
- Fixed `library.json` to point the repo url to TRT instead of my account (holdover from my fork and initial setup of that file).
- Added Abhi and Avi as authors to `library.json`.
- Removed KF repo from list of dependencies.
- Removed duplicate Windows function definitions in `BlinkBuzz.h`.
- Added several more constants designed to help unify time lengths to `Constants.h`.
- Removed `UNKNOWN_` type from SensorType and ActuatorType.
- Added a `+=` operator to the vector class.
- Added more header files to `MMFS.h`.
- Changed APRS message types to be `const`, as they are string literals.
- Added a proper `double` APRS  base91 encoder to prevent int rounding or truncation.
- Fixed `decodeBase91()` call for altitude attempting to decode only 2 of the three digits.
- Sensors can now be fed a char literal for their name in the constructor, with a default to the name of the sensor.
- Hardware level updating now occurs through a `read()` (not `update()` anymore) function, where ONLY the hardware reading is done. This function should almost always be called ONLY internally by `SensorType.update()`. More documentation to come later.
- Slightly modified BMP390 initialization to ensure all commands were properly registered.
- Fixed faulty altitude determination that was mistakenly using measured ground pressure as current sea level pressure.
- Added `getASLAlt()` and `getAGLAlt()` to barometer to differentiate between absolute altitude and the delta from power on (or launch) to now.
- Added both ASL and AGL (Above Sea Level and Above Ground Level) fields to the Barometer data string

# Major Notes:

The IMU implementation in this version is currently... well... not implemented. Changes to IMU hardware brought up the fact that we need to add math to get global frame from the IMU's local frame. I can do this math, but I feel more discussion needs to happen on wwhat this looks like. I think Ezra would say to not even worry about it because you can just put the raw values into the KF and still get the values you need (at least that's my understanding, but I admittedly don't understand how Kalman Filters work).

The `State` tests are currently not implemented as the IMU is still not implemented. Once all sensors are done, state tests can begin.

I will write documentation about how to both use and modify this code as I begin migrating the Avionics wiki over here. I think I may slightly change the layout so that the wiki is split into two broad categories: using the software and modifying the software. One goes over the API and the other dives into the internals to help people understand how it currently works so that they can make changes if necessary.

# Updates after KF and RecordData updates:
- Changed log types to include `_` after the name (`ERROR_`, etc) because Windows included `ERROR` as a macro and I couldn't find a good way to undefine that macro.
- Added Arduino.h to `SDCardStorage`, not sure how it built before without that there.
- Changed the sensor `read()` and `init()` functions to `protected`
- Removed logging data from GPS for now (will become a new issue)
- Changed the BNO055's `update()` to `read()` like it's supposed to be

varun-un commented:
Should the `Math` folder be renamed to something like `Utils` if it now has utilities like the circular buffer

varun-un commented on src/Math/CircBuffer.h:
```diff
@@ -0,0 +1,112 @@
+    T *buffer;
+    int size;
+    int head;
+    int tail;
```
Do you need to initialize these to 0?

varun-un commented on src/Math/CircBuffer.h:
```diff
@@ -0,0 +1,112 @@
+#ifndef CIRC_BUFFER_H
+#define CIRC_BUFFER_H
+
```
Add a file level comment explaining generically how to use it and what overrides are there (like bracket notation accessing)

varun-un reviewed (commented):
Before we merge I think we should have some documentation attached about how to use the unit tests somewhere, and how to setup Unity

If it doesn't build with the Arduino IDE, we may want to wait to merge it, but if the "fake" environment you mentioned works to at least avoid build errors, then it should be fine.

jhauerst commented on examples/AvionicsState.cpp:
Why was this code left commented out?

jhauerst commented on lib/NativeTestMocks/Arduino.h:
```diff
@@ -0,0 +1,37 @@
+void digitalWrite(int pin, int value);
+
+class SerialClass {
+    public:
```
It might be necessary to have a Serial.begin() here as well if we're trying to create a mock Serial.

jhauerst commented on src/Math/Matrix.h:
```diff
@@ -0,0 +1,43 @@
+#include <iostream>
+#include <cmath>
```
Why are standard C/C++ libraries included here? If I remember correctly we decided to try to stay away from them for the Teensy even though they are supported, but iostream definitely won't work on a Teensy.

DrewBrandt replied:
Left over from porting from KF repo

jhauerst commented on src/Math/Quaternion.h:
```diff
@@ -0,0 +1,216 @@
+  //     _z = 0.25 * S;
+  //   }
+  // }
+
```
The commented code here should probably either be removed or uncommented.

DrewBrandt replied:
I'd like to keep it around in case we want to use it in the future. It's supposed to use the imu matrix class, not our own, which is why I commented it out. The rest of the class is verbatim Adafruit's Quaternion class and I didn't want to change it. The code should ideally not ever need to be modified, so leaving the comments there shouldn't be a big deal. I will add another comment explaining why it's commented.

jhauerst replied:
Ok, sounds good to me. Although if we don't end up using it for a while we should probably revisit removing just to keep the code cleaner.

jhauerst commented on src/Sensors/Baro/BMP280.cpp:
```diff
@@ -3,40 +3,31 @@
             return initialized = false;
         }
 
-        delay(1000);
```
Was there a reason this delay was in here before?

DrewBrandt replied:
Most likely to help the sensor stabilize, but by the time this function is run, the board will have been powered for multiple seconds.

jhauerst commented on src/Sensors/Baro/BMP390.cpp:
```diff
@@ -3,46 +3,37 @@
         }
 
-        delay(1000);
+        // delay(1000);
```
Same question as BMP280 about the delay

DrewBrandt replied:
Most likely to help the sensor stabilize, but by the time this function is run, the board will have been powered for multiple seconds.

jhauerst commented on src/Sensors/Baro/BMP390.cpp:
```diff
@@ -3,46 +3,37 @@
-            bmp.readPressure();
-            delay(25);
-        }
-        for (int i = 0; i < 100; i++)
```
Why was the first for loop removed for the 390 but not the 280? Did the initialization procedures end up being different?

DrewBrandt replied:
oversight

jhauerst commented on src/Sensors/GPS/GPS.h:
```diff
@@ -1,46 +1,47 @@
     public:
         virtual ~GPS(){};
-        virtual double getAlt() const;
-        virtual imu::Vector<3> getVelocity() const;
```
Even though (if I'm remembering right) the GPS velocity wasn't completely accurate, do we want to completely remove the ability to get velocity from the GPS?

DrewBrandt replied:
The problem here was not that it was inaccurate, but that the GPS sensor was never providing it, and so it doesn't really belong here. The velocity was being calculated by us manually. Not some internal sensor algorithm. If a user wants specifically the GPS velocity, I believe it was decided that it would be better for them to do it themselves in State.

jhauerst replied:
That works for me

jhauerst commented on src/Sensors/GPS/MAX_M10S.cpp:
```diff
@@ -18,7 +18,8 @@ namespace mmfs
         {
             // Serial.println(F("u-blox GNSS not detected at default I2C address. Retrying..."));
-            delay(1000);
+            // TODO: RECORD FAILURE IN RECORDDATA
```
Should probably create a issue to address this or add it to an existing issue

jhauerst commented on src/Sensors/IMU/IMU.h:
```diff
@@ -2,38 +2,64 @@
+    // - Implement Magnetometer reading fusion
+    // - Revist naming conventions
+    // - Revisit necessary funcitons/variables
+
```
Also should create an issue for this

jhauerst commented on src/Sensors/Sensor.h:
```diff
@@ -1,8 +1,13 @@
+#include "../Math/CircBuffer.h"
+#include "../BlinkBuzz/BlinkBuzz.h"
+#include <string.h>
+#include <stdio.h>
```
What are string.h and stdio.h being used for? I'm pretty sure Arduino.h should provide most of the functions we're using from them.

jhauerst commented on src/State/State.cpp:
```diff
@@ -73,17 +73,17 @@ namespace mmfs
+                //     sensors[i]->update();
+                //     delay(10);
+                //     sensors[i]->update();
+                // }
```
Why was this check commented out?

DrewBrandt replied:
A few reasons: 
- It is only there because the BNO055 specifically sucks
- We were reworking the record data library
- I would have had to write a `Wire` stub and this was easier. 

I can probably uncomment it if we think we need it. probably depends on whether or not anyone plans to use the BNO055. Maybe it should be specifically added to the BNO055 update method instead of here.

DrewBrandt replied:
Made an issue for it

jhauerst commented:
> Before we merge I think we should have some documentation attached about how to use the unit tests somewhere, and how to setup Unity
> 
> If it doesn't build with the Arduino IDE, we may want to wait to merge it, but if the "fake" environment you mentioned works to at least avoid build errors, then it should be fine.

I agree that we should have at least some documentation on how to use the tests before this is merged. It doesn't have to be anything super detailed, just some basic steps for running the tests so that people will know how to use them.

mallamacimj commented:
> Should the `Math` folder be renamed to something like `Utils` if it now has utilities like the circular buffer

IMO we should move the circular buffer to a Utils folder but keep all the math in a math folder. I think that will be the most logical to anyone new to the code.

DrewBrandt commented:
# Updates after KF and RecordData updates:
- Changed log types to include `_` after the name (`ERROR_`, etc) because Windows included `ERROR` as a macro and I couldn't find a good way to undefine that macro.
- Added Arduino.h to `SDCardStorage`, not sure how it built before without that there.
- Changed the sensor `read()` and `init()` functions to `protected`
- Removed logging data from GPS for now (will become a new issue)
- Changed the BNO055's `update()` to `read()` like it's supposed to be

DrewBrandt commented on lib/NativeTestMocks/Arduino.h:
Potentially, yes, but I did this because it was easy to do and I could very easily understand how to use it and how to modify it. Some of the other mock frameworks I looked at seemed much more complicated to use and didn't have great documentation. With this, we simply specify to build or not build this folder when on or not on native, and it does the rest. (It's already set up that way so the only real maintenance is if we start calling more methods that this doesn't mock, then we have to add those.

DrewBrandt commented on src/RecordData/Logger.cpp:
```diff
@@ -74,7 +74,7 @@ void Logger::recordLogData(LogType type, const char *data, Dest dest) {
 
 // Records log data with a specific timestamp, type, and destination
-void Logger::recordLogData(double timeStamp, LogType type, const char *data, Dest dest = BOTH) {
+void Logger::recordLogData(double timeStamp, LogType type, const char *data, Dest dest) {
```
The default here is redundant because it's already given a specified default in the .h file. Having two default declarations generated a warning so I removed it here.

DrewBrandt commented on src/Sensors/Baro/BMP390.cpp:
```diff
@@ -3,46 +3,37 @@
         }
 
-        delay(1000);
+        // delay(1000);
```
Incase we found that it was necessary mostly. Didn't want to remove potentiaally necessary blocking calls before we had a chance to verify that such calls were overkill. At this point, the board will be powered on for a while before it gets here (like 2-3 seconds) so I felt this wait was unnecessary and only served to push the main loop start point back. But, if, after testing, we notice unwanted behavior we can easily come back in and see hte comment and say "oh I wonder if this commented code has something to do with it" and test that to see if it helps.

jhauerst commented:
Docs look good to me and tests ran without issue

varun-un commented on src/Math/Matrix.cpp:
We could, but then we restrict Math library to be dependent on Logger, which idk if that's a paradigm we want especially if Logger is meant to be higher level.

The way around this would be to return an Error object, but with class types idk how that would work

varun-un commented on test/test_gps/TestGPS.cpp:
```diff
@@ -0,0 +1,269 @@
+constexpr double lat2 = 38.8977;
+constexpr double lon2 = -77.0365; // Washington, DC
+
+// Just gotta say its freaky af that CoPilot knew those were CP MD coords, then predicted that I would want DC coords next (and gave me the correct coords). Weird af.
```
real

varun-un reviewed (approved):
lgtm
